File: app/src/main/java/com/damianogiusti/cleanandroid/di/modules/MvpModule.java
package com.damianogiusti.cleanandroid.di.modules;

import com.damianogiusti.cleanandroid.mvp.PresenterProvider;
import com.damianogiusti.cleanandroid.ui.home.MainPresenter;
import com.damianogiusti.domain.cleanandroid.interactors.GetProvinceListUseCase;

import dagger.Module;
import dagger.Provides;

/**
 * Created by Damiano Giusti on 03/05/17.
 */
@Module
public class MvpModule {

    private PresenterProvider presenterProvider;

    public MvpModule(PresenterProvider presenterProvider) {
        this.presenterProvider = presenterProvider;
    }

    @Provides
    MainPresenter providesMainPresenter(GetProvinceListUseCase getProvinceListUseCase) {
        MainPresenter presenter = (MainPresenter) presenterProvider.getRetainedPresenter();
        if (presenter == null) {
            presenter = new MainPresenter(getProvinceListUseCase);
        }
        return presenter;
    }
}

File: app/src/main/java/com/damianogiusti/cleanandroid/di/modules/MvpModule.kt
package com.damianogiusti.cleanandroid.di.modules

import com.damianogiusti.cleanandroid.mvp.PresenterProvider
import com.damianogiusti.cleanandroid.ui.home.MainPresenter
import com.damianogiusti.domain.cleanandroid.interactors.GetProvinceListUseCase

import dagger.Module
import dagger.Provides

/**
 * Created by Damiano Giusti on 03/05/17.
 */
@Module
class MvpModule(private val presenterProvider: PresenterProvider) {

    @Provides
    internal fun providesMainPresenter(getProvinceListUseCase: GetProvinceListUseCase): MainPresenter {
        var presenter = presenterProvider.getRetainedPresenter() as MainPresenter?
        if (presenter == null) {
            presenter = MainPresenter(getProvinceListUseCase)
        }
        return presenter
    }
}
